compiler/src/iree/compiler/Codegen/Common/GPU/GPUPropagateDispatchSizeBounds.cpp

```diff
@@ -0,0 +1,87 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Common/GPU/Passes.h"
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h"
+#include "iree/compiler/Codegen/Utils/Utils.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+#include "mlir/Transforms/Passes.h"
+
+#define DEBUG_TYPE "iree-codegen-gpu-propagate-dispatch-size-bounds"
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_GPUPROPAGATEDISPATCHSIZEBOUNDSPASS
+#include "iree/compiler/Codegen/Common/GPU/Passes.h.inc"
+
+namespace {
+
+static void applyBounds(FunctionOpInterface funcOp,
+                        ArrayRef<int64_t> workitemCounts,
+                        ArrayRef<int64_t> workgroupCounts) {
+  Builder b(funcOp->getContext());
+  funcOp->walk([&](Operation *op) {
+    if (auto tidOp = dyn_cast<gpu::ThreadIdOp>(op)) {
+      // Not the actual code, pending an upstream merge.
+      tidOp.setUpperBoundAttr(b.getIndexAttr(
+          workitemCounts[static_cast<uint32_t>(tidOp.getDimension())]));
+    } else if (auto wgSizeOp =
+                   dyn_cast<IREE::HAL::InterfaceWorkgroupSizeOp>(op)) {
+      wgSizeOp.setUpperBoundAttr(b.getIndexAttr(
+          workitemCounts[wgSizeOp.getDimension().getZExtValue()]));
+    } else if (auto wgIdOp = dyn_cast<IREE::HAL::InterfaceWorkgroupIDOp>(op)) {
+      wgIdOp.setUpperBoundAttr(b.getIndexAttr(
+          workgroupCounts[wgIdOp.getDimension().getZExtValue()]));
+    } else if (auto wgCountOp =
+                   dyn_cast<IREE::HAL::InterfaceWorkgroupCountOp>(op)) {
+      wgCountOp.setUpperBoundAttr(b.getIndexAttr(
+          workgroupCounts[wgCountOp.getDimension().getZExtValue()]));
+    }
+  });
+}
+
+struct GPUPropagateDispatchSizeBoundsPass final
+    : impl::GPUPropagateDispatchSizeBoundsPassBase<
+          GPUPropagateDispatchSizeBoundsPass> {
+  using GPUPropagateDispatchSizeBoundsPassBase::
+      GPUPropagateDispatchSizeBoundsPassBase;
+  void runOnOperation() override {
+    FunctionOpInterface funcOp = getOperation();
+    SmallVector<int64_t> workitemCounts(3, maxWorkgroupSize);
+    SmallVector<int64_t> workgroupCounts = {maxWorkgroupsX, maxWorkgroupsYZ,
+                                            maxWorkgroupsYZ};
+
+    std::optional<SmallVector<int64_t>> staticWorkitemCounts =
+        getWorkgroupSize(funcOp);
+    if (staticWorkitemCounts) {
+      if (staticWorkitemCounts->size() > 3) {
+        // How to flag this but not fail the pipeline?
+        LLVM_DEBUG(llvm::dbgs()
+                   << "more than 3 workgroup size dimensions in GPU code\n");
```

Comment:
`funcOp.emitWarning`

Reply:
Ah, that's the right style, thanks, will fix